File: api/Controllers/TransacaoController.cs
using api.DAO;
using api.Decorators;
using api.Models;
using api.Repositories;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using MySqlX.XDevAPI;

namespace api.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class TransacaoController : ControllerBase
    {
        private readonly IClienteRepository _clienteDAO;

        public TransacaoController()
        {
            _clienteDAO = new ClienteDAO();
        }

        [Route("deposito")]
        [HttpPost]
        public ActionResult<Transacao> Depositar([FromForm]int idCliente, [FromForm]double valor)
        {
            TransacaoDados dados = new() 
            {
                IdCliente = idCliente,
                Valor = valor,
            };

            var _transacaoDAO = new TransacaoDAO(TipoTransacao.deposito);
            var transacaoResponse = _transacaoDAO.ExecutarTransacao(dados);

            if (transacaoResponse == null) return BadRequest(); 

            return Ok(transacaoResponse);
        }

        [Route("saque")]
        [HttpPost]
        public ActionResult<Transacao> Sacar([FromForm] int idCliente, [FromForm] double valor)
        {
            TransacaoDados dados = new()
            {
                IdCliente = idCliente,
                Valor = valor,
            };

            var _transacaoDAO = new TransacaoDAO(TipoTransacao.saque);
            var _transacaoDecorator = new VerificarSaldoDecorator(_transacaoDAO, _clienteDAO);
            var transacaoResponse = _transacaoDecorator.ExecutarTransacao(dados);

            if (transacaoResponse == null) return BadRequest();

            return Ok(transacaoResponse);
        }

        [Route("transferencia")]
        [HttpPost]
        public ActionResult<Transacao> Transferir([FromForm] int idCliente, [FromForm] int idClienteTransf, [FromForm] double valor)
        {
            TransacaoDados dados = new()
            {
                IdCliente = idCliente,
                IdClienteTransf = idClienteTransf,
                Valor = valor,
            };

            var _transacaoDAO = new TransacaoDAO(TipoTransacao.tranferencia);
            var _transacaoDecorator = new VerificarSaldoDecorator(_transacaoDAO, _clienteDAO);
            var transacaoResponse = _transacaoDecorator.ExecutarTransacao(dados);

            if (transacaoResponse == null) return BadRequest();

            return Ok(transacaoResponse);
        }

        [Route("buscar")]
        [HttpGet]
        public ActionResult<List<TransacaoRegistro>> BuscarTransacoes([FromQuery] int idCliente)
        {
            var _transacaoTransf = new TransacaoDAO(TipoTransacao.tranferencia);
            var _transacaoSaque = new TransacaoDAO(TipoTransacao.saque);
            var _transacaoDeposito = new TransacaoDAO(TipoTransacao.deposito);

            List<List<TransacaoRegistro>?> responses = new()
            { 
                _transacaoTransf.BuscarTransacoesCliente(idCliente),
                _transacaoSaque.BuscarTransacoesCliente(idCliente),
                _transacaoDeposito.BuscarTransacoesCliente(idCliente)
            };
            

            List<TransacaoRegistro> registros = new();
            responses.ForEach(response =>
            {
                if (response != null)
                    registros.AddRange(response);
            });

            if (registros == null) return BadRequest();

            return Ok(registros);
        }
    }
}

File: api_node/src/controllers/transacao-controller.ts
import { Request, Response } from 'express';
import { TipoTransacao } from '../enums/tipo-transacao-enum';
import { TransacaoDAO } from '../dao/transacao-dao';
import { ValidadeTransacaoDecorator } from '../decorators/validate-transacao-decorator';
import { ClienteDAO } from '../dao/cliente-dao';

export class TransacaoController {

  public async depositar(req: Request, res: Response) {
    const transacaoRepository = new TransacaoDAO(TipoTransacao.deposito);
    try {
      const { ...data } = req.body;
      const resposta = await transacaoRepository.executarTransacao({ 
        tipo: TipoTransacao.deposito,
        ...data
      });
      res.status(200).json(resposta);
    }
    catch(error) {
      res.status(400).json({ error });
    }
  }

  public async sacar(req: Request, res: Response) {
    const transacaoRepository = new TransacaoDAO(TipoTransacao.saque);
    const clienteRepository = new ClienteDAO();
    const transacaoDecorator = new ValidadeTransacaoDecorator(
      transacaoRepository, 
      clienteRepository
    );
    try {
      const { ...data } = req.body;
      const resposta = await transacaoDecorator.executarTransacao({ 
        tipo: TipoTransacao.saque,
        ...data
      });

      if(resposta)
        return res.status(200).json(resposta);

      return res.status(400).send('saldo insuficiente');
    }
    catch(error) {
      res.status(400).json({ error });
    }
  }

  public async transferir(req: Request, res: Response) {
    const transacaoRepository = new TransacaoDAO(TipoTransacao.transferencia);
    const clienteRepository = new ClienteDAO();
    const transacaoDecorator = new ValidadeTransacaoDecorator(
      transacaoRepository, 
      clienteRepository
    );
    try {
      const { ...data } = req.body;
      const resposta = await transacaoDecorator.executarTransacao({ 
        tipo: TipoTransacao.transferencia,
        ...data
      });
      
      if(resposta)
        return res.status(200).json(resposta);

      return res.status(400).send('saldo insuficiente');
    }
    catch(error) {
      res.status(400).json({ error });
    }
  }

  public async buscar(req: Request, res: Response) {
    const transacaoSaque = new TransacaoDAO(TipoTransacao.saque);
    const transacaoDeposito = new TransacaoDAO(TipoTransacao.deposito);
    const transacaoTransferencia = new TransacaoDAO(TipoTransacao.transferencia);
    try {
      const idCliente = req.query.idCliente as string;

      const [resSaque, resDeposito, resTransferencia] = await Promise.all([
        await transacaoSaque.buscarTransacoes(parseInt(idCliente)),
        await transacaoDeposito.buscarTransacoes(parseInt(idCliente)),
        await transacaoTransferencia.buscarTransacoes(parseInt(idCliente))
      ]);

      const respostas = [resSaque?? [], resDeposito?? [], resTransferencia?? []];
      const registros = ValidadeTransacaoDecorator.ordenarTrasacoesByData(respostas);

      res.status(200).json(registros);
    }
    catch(error) {
      res.status(400).json({ error });
    }
  }
}